### ISSUE TITLE
Recursive property pattern allows reference to same property multiple times.

### ISSUE THREAD

**Issue by CyrusNajmabadi** [dotnet/csharplang#4120]
Example:

```c#
void M(string s)
{
    if (s is { Length: 0, Length: var v })
    {
    }
}
```

This honestly feels strange to me and potentially a bug oversight.  If a bug/oversight, not sure if it can be fixed due to the breaking chnage.  However, we should def look into this to determine if this is intentional/desirable and at least doc appropriately.

**Comment by CyrusNajmabadi** [comment]
transferred to csharplang and reopened.  I think it would be good to track adding info that is explicit about this being allowed.

**Comment by alrz** [comment]
Can we make this a warning wave at least? This doesn't quite feel right to accept as an idiomatic valid C# code, even the behavior is correct.

**Comment by jnm2** [comment]
TIL that C# 8 had `and` patterns on properties :D

**Comment by alrz** [comment]
Yeah but when I see `{ prop: pattern, ... }` I don't want to expect even more checks on that property and certainly not something we should keep not warn about as if it's *fine*. Definitely a candidate for waning waves IMO.

**Comment by HaloFour** [comment]
Maybe just an analyzer with a fixer to suggest a conjunctive pattern instead?  I can see people preferring two property patterns instead of a conjunctive pattern in some cases.

**Comment by jaredpar** [comment]
@alrz  

> Can we make this a warning wave at least? 

I wouldn't support a warning wave here. There is nothing wrong about doing this. The difference between `{ Length: > 0, Length: < 42 }` and `{ Length: > 0 and < 42 }` is purely stylistic. Warnings should be reserved for cases where it is very likely that we are pushing users toward the pit of failure here. That does not seem to be the case here.

**Comment by jaredpar** [comment]
@CyrusNajmabadi 

>  I think it would be good to track adding info that is explicit about this being allowed.

We don't explicitly state that the same porperty can be used with operators like `&&`. Why should we make it explicit here? 

Or think of this a different way. Pretty much every sub-expression allows for the same method, field, property, etc ... to be re-used in various arms. Why are patters special here that we have to call out this is legal?

**Comment by alrz** [comment]
The reason I think this is a bad code pattern is that patterns are generally a declarative way to express some condition to be satisfied without repeating the target, for instance in a range check like `x > 1 && x < 10`, the `x` is important to be the same in both sides otherwise either it's a bug or it's not a range check anymore. We can write `x is > 1 and < 10` and then there's no room for error.

When I see `{ Length: 0, .. }` I want it to be the only thing checking length so I don't have to scan the whole pattern looking for other things that might affect it, immediately losing the advantage of using a pattern in the first place.

I think a warning is appropriate because I believe it should've been an error.

**Comment by jaredpar** [comment]
Errors are reserved for code that cannot be correctly executed or if when it was executed would produce results countertuitive to what the user wrote. This case does not present this problem. The behavior has well defined meaning. The fact that the user chose to repeat properties vs. using an explicit `and` is a style decision.

**Comment by alrz** [comment]
>Errors are reserved for code that cannot be correctly executed or if when it was executed would produce results countertuitive to what the user wrote. This case does not present this problem. The behavior has well defined meaning. The fact that the user chose to repeat properties vs. using an explicit and is a style decision.

Syntax can and does enforce limitation for neither of those reasons. For instance, does that apply to object initializers? Before initonly, yes it was only a code style preference. Reassignments can be correctly executed and possibly intended for side effects. But the fact that object initializers only allow one assignment per member adds a guarantee that helps to prevent bugs and improve readability.

**Comment by jaredpar** [comment]
Object initializers are a case where multiple uses of a property is specifically wrong. It is demonstrably calculating values only to throw them away. 

Using the same property twice in a pattern does not have that behavior. The property value is used and is used in a well defined way consistent with other properties in the pattern. There is nothing wrong about this, it is just a style decision on how the values are used.

**Comment by alrz** [comment]
>Object initializers are a case where multiple uses of a property is specifically wrong. It is demonstrably calculating values only to throw them away.

Not if I have a custom setter that do anything but throwing away the value. 

In both cases, repeating the property is in fact completely valid and very well may be the user intention, but we chose to reject the former for a good reason of which I believe applies to patterns as well.

We're clearly not going to agree on that but if it is seen as a code style preference it would also make sense to have a code style option as well to prefer `and` vs. repeated props or vice versa.

**Comment by jaredpar** [comment]
> In both cases, repeating the property is in fact completely valid and very well may be the user intention, but we chose to reject the former for a good reason of which I believe applies to patterns as well.

Disagree. In the object initializer case for the mulitple property case to be valid you have to effectively have an anti-propery. A property that exists specifically to discard values. Whereas with the pattern case completely valid properties work in completely valid ways here.

**Comment by alrz** [comment]
>you have to effectively have an anti-propery

Exactly. In the language, a specific syntax could be more than just a syntactic sugar to avoid duplicated code, like discouraging code smells and adding some guarantees to enforce DRY and reducing cognitive load. But as far as the compiler concerns, at the end there's nothing wrong to repeat all that in either patterns, or object initializers to some extent.

**Comment by jaredpar** [comment]
> But as far as the compiler concerns, at the end there's nothing wrong to repeat all that in either patterns, or object initializers to some extent.

Disagree. The compiler makes assumptions in several places that properties are well behaving: object intiliazers, patterns, struct based setters and nullable reference types to name a few. The fact that we issue an error for object initializers but not for property patterns is completely consistent with this.

**Comment by alrz** [comment]
>The fact that we issue an error for object initializers but not for property patterns is completely consistent with this.

Makes sense. I wish I had filed a bug for this and I very much hope that it would have been closed as by-design.

**Comment by jaredpar** [comment]
It's also fair to argue whether the compiler should or shouldn't make assumptions about how properties behave. This has been the topic of much debate before. 

I do think the tide has turned a bit in more recent times towards making assumptions about how properties behave: particularly that they are side effect free. Two recent features that are pretty much predicated on this being true are patterns and nullable reference types. Those features would've been much different, if possible at all, if we had kept a hard line on not making assumptions about property side effects.

**Comment by alrz** [comment]
I understand that this is a byproduct of those assumptions in the implementation and that's why it doesn't incur any invalid behavior as far as those assumptions are held. But syntax-wise I still think this would have been considered a bug at the time, though not something we need to *fix* right now.

**Comment by CyrusNajmabadi** [comment]
> Why should we make it explicit here?

Because being explicit removes any potential confusion.

> We don't explicitly state that the same porperty can be used with operators like &&

Right.  But we do disallow doing things like `new Foo { X = 1, X = 2 }`, or `foo with { X = 1, X = 2 }` and it's reasonable to ahve an intuition that when doing property patterns (like with property initialization) that the member can only be mentioned once.  

As I've said, i'm fine with that not being hte case.  I just think it would be good to explicitly doc that as i think more information about something unintuitive is good.  

--

Note: as an experiment, i checked with two other team members and asked if they thought this code would be legal.  They did not.  So we had three people on the team itself that were surpised by this.  When the team that is creating the language finds something surprising, then it's a good time to doc it :)

**Comment by rootflood** [comment]
hi, im not professional but i like this one
```c#
void M(string s)
{
    if (s.Length is {  value > 0, value < 42 })
    {
    }
}
```

my reason is performance, this mean s.Length (Get Method) run just once
on this one the length property is not heavy operation but if we have heavy one this can work faster
in other hand i don't want see extra variable in the flowing code
currently i writing something like below :
```c#
void M(string s)
{

    {
        var value = s.Length;
        if (value > 0, value < 42 )
        {
        }
    }
    
    var value = s + "str";
}
```
This is just my personal opinion ...

**Comment by jnm2** [comment]
There's a currently use case for repeated property patterns:

```cs
if (expr is
{
    Foo: { Bar: 1 } or { Baz: 2},
    Foo: { Instance: ISomethingSpecific instance },
})
{
    instance.Foo();
}  
```

**Comment by jnm2** [comment]
Never mind, missed the obvious:

```cs
if (expr is
{
    Foo: ({ Bar: 1 } or { Baz: 2 }) and { Instance: ISomethingSpecific instance },
})
{
    instance.DoStuff();
}  
```
